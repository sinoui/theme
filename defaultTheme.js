module.exports = require('./dist/index').defaultTheme;
