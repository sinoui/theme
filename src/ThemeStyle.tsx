import { createGlobalStyle, css } from 'styled-components';
import { ThemeTypography } from './types';

const typographyVariants: [keyof ThemeTypography, string][] = [
  ['h1', 'h1'],
  ['h2', 'h2'],
  ['h3', 'h3'],
  ['h4', 'h4'],
  ['h5', 'h5'],
  ['h6', 'h6'],
  ['subtitle1', '.subtitle1'],
  ['subtitle2', '.subtitle2'],
  ['body1', '.body1'],
  ['body2', '.body2'],
  ['caption', '.caption'],
  ['overline', '.overline'],
];

/**
 * 文字排版样式
 */
const typographyStyles = css(({ theme }) =>
  typographyVariants.reduce((acc, [key, cssSelector]) => {
    acc[cssSelector] = theme.typography[key];
    return acc;
  }, {} as Record<string, any>),
);

/**
 * 链接样式
 */
const linkStyle = css`
  a {
    color: ${(props) => props.theme.palette.secondary.main};
    text-decoration: none;
  }

  a:hover {
    text-decoration: underline;
  }
`;

interface Props {
  /**
   * 如果设置为 `true`，则不会产生文字排版相关的css。
   */
  disableTypography?: boolean;
  /**
   * 如果设置为 `true`，则不会为链接添加默认样式。
   */
  disableLink?: boolean;
}

/**
 * 主题全局样式
 */
const ThemeStyle = createGlobalStyle<Props>`
  html {
    font-size: ${(props) => props.theme.typography.htmlFontSize}px;
  }
  body {
    font-family: ${(props) => props.theme.typography.fontFamily};
    font-size: ${(props) => props.theme.typography.body2.fontSize};
    color: ${(props) => props.theme.palette.text.primary};
    font-variant: tabular-nums;
    background-color: ${(props) => props.theme.palette.background.default};
  }

  ${({ disableTypography }) => !disableTypography && typographyStyles}

  ${({ disableLink }) => !disableLink && linkStyle}
`;

export default ThemeStyle;
