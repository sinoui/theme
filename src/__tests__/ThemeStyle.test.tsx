/* eslint-disable class-methods-use-this */
import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider, StyleSheetManager } from 'styled-components';
import 'jest-styled-components';
import ThemeStyle from '../ThemeStyle';
import defaultTheme from '../defaultTheme';

it('产生全局样式', () => {
  const inject = jest.fn();
  renderer.create(
    <StyleSheetManager
      sheet={
        {
          hasId() {
            return false;
          },
          remove: () => 0,
          inject,
        } as any
      }
    >
      <ThemeProvider theme={defaultTheme}>
        <ThemeStyle />
      </ThemeProvider>
    </StyleSheetManager>,
  );

  expect(inject.mock.calls[0][1].join('\n')).toMatchSnapshot();
});

it('不产生字体分类样式', () => {
  const inject = jest.fn();
  renderer.create(
    <StyleSheetManager
      sheet={
        {
          hasId() {
            return false;
          },
          remove: () => 0,
          inject,
        } as any
      }
    >
      <ThemeProvider theme={defaultTheme}>
        <ThemeStyle disableTypography />
      </ThemeProvider>
    </StyleSheetManager>,
  );

  expect(inject.mock.calls[0][1].join('\n')).toMatchSnapshot();
});
