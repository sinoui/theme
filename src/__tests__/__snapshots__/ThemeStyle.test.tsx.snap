// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`不产生字体分类样式 1`] = `
"html{font-size:16px;}
body{font-family:-apple-system,BlinkMacSystemFont,'Segoe UI','PingFang SC','Hiragino Sans GB','Microsoft YaHei',Roboto,'Noto Sans SC','Helvetica Neue',Helvetica,Arial,sans-serif,'Apple Color Emoji','Segoe UI Emoji','Segoe UI Symbol';font-size:0.875rem;color:rgba(0,0,0,0.87);font-variant:tabular-nums;background-color:#fafafa;}
a{color:#f50057;-webkit-text-decoration:none;text-decoration:none;}
a:hover{-webkit-text-decoration:underline;text-decoration:underline;}"
`;

exports[`产生全局样式 1`] = `
"html{font-size:16px;}
body{font-family:-apple-system,BlinkMacSystemFont,'Segoe UI','PingFang SC','Hiragino Sans GB','Microsoft YaHei',Roboto,'Noto Sans SC','Helvetica Neue',Helvetica,Arial,sans-serif,'Apple Color Emoji','Segoe UI Emoji','Segoe UI Symbol';font-size:0.875rem;color:rgba(0,0,0,0.87);font-variant:tabular-nums;background-color:#fafafa;}
h1{font-weight:300;font-size:6rem;line-height:1.167;-webkit-letter-spacing:-0.01562em;-moz-letter-spacing:-0.01562em;-ms-letter-spacing:-0.01562em;letter-spacing:-0.01562em;}
h2{font-weight:300;font-size:3.75rem;line-height:1.2;-webkit-letter-spacing:-0.00833em;-moz-letter-spacing:-0.00833em;-ms-letter-spacing:-0.00833em;letter-spacing:-0.00833em;}
h3{font-weight:400;font-size:3rem;line-height:1.167;-webkit-letter-spacing:0em;-moz-letter-spacing:0em;-ms-letter-spacing:0em;letter-spacing:0em;}
h4{font-weight:400;font-size:2.125rem;line-height:1.235;-webkit-letter-spacing:0.00735em;-moz-letter-spacing:0.00735em;-ms-letter-spacing:0.00735em;letter-spacing:0.00735em;}
h5{font-weight:400;font-size:1.5rem;line-height:1.334;-webkit-letter-spacing:0em;-moz-letter-spacing:0em;-ms-letter-spacing:0em;letter-spacing:0em;}
h6{font-weight:500;font-size:1.25rem;line-height:1.6;-webkit-letter-spacing:0.0075em;-moz-letter-spacing:0.0075em;-ms-letter-spacing:0.0075em;letter-spacing:0.0075em;}
.subtitle1{font-weight:400;font-size:1rem;line-height:1.75;-webkit-letter-spacing:0.00938em;-moz-letter-spacing:0.00938em;-ms-letter-spacing:0.00938em;letter-spacing:0.00938em;}
.subtitle2{font-weight:500;font-size:0.875rem;line-height:1.57;-webkit-letter-spacing:0.00714em;-moz-letter-spacing:0.00714em;-ms-letter-spacing:0.00714em;letter-spacing:0.00714em;}
.body1{font-weight:400;font-size:1rem;line-height:1.5;-webkit-letter-spacing:0.00938em;-moz-letter-spacing:0.00938em;-ms-letter-spacing:0.00938em;letter-spacing:0.00938em;}
.body2{font-weight:400;font-size:0.875rem;line-height:1.43;-webkit-letter-spacing:0.01071em;-moz-letter-spacing:0.01071em;-ms-letter-spacing:0.01071em;letter-spacing:0.01071em;}
.caption{font-weight:400;font-size:0.75rem;line-height:1.66;-webkit-letter-spacing:0.03333em;-moz-letter-spacing:0.03333em;-ms-letter-spacing:0.03333em;letter-spacing:0.03333em;}
.overline{font-weight:400;font-size:0.75rem;line-height:2.66;-webkit-letter-spacing:0.08333em;-moz-letter-spacing:0.08333em;-ms-letter-spacing:0.08333em;letter-spacing:0.08333em;text-transform:uppercase;}
a{color:#f50057;-webkit-text-decoration:none;text-decoration:none;}
a:hover{-webkit-text-decoration:underline;text-decoration:underline;}"
`;
